# **(Data collection Notebook)**


## 

Objectives
Fetch data from Kaggle and sit ave as raw data
Inspect the data and save it under outputs/datasets/collect
## ion.
Inputs
Kaggle JSON file - the authentication token.
Raw downloaded da
## ta file
Outputs
Generate and save dataset: outputs/dataset/c etction/ ....

rn.csv
AdIn the workplace, projects are not done using Kaggle data, but instead, the data comes from multiple data sources that may be hosted internally (like in a data warehouse) or outside your company. For this project learning context, we are fetching the data from Kaggle.

Another aspect is that in the workplace, the data has never been pushed to a public repository due to security reasons. Just for this project learning context, we are hosting the data in a public repo.a in a publhem here. 


---

## Install python packages in the notebooks

In [1]:
%pip install -r /workspaces/house_sales_predictor/requirements.txt

Note: you may need to restart the kernel to use updated packages.


# Change working directory

* Notebooks are stored in a subfolder, therefore when running the notebook in the editor, we need to change the working directory from the current folder to the parent folder.

* We access the current directory with os.getcwd()

In [2]:
import os
current_dir = os.getcwd()
current_dir

'/workspaces/house_sales_predictor/jupyter_notebooks'

## We want to make the parent of the current directory the new current directory

* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [3]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


## Confirm the new current directory

In [4]:
current_dir = os.getcwd()
current_dir

'/workspaces/house_sales_predictor'

# Fetch data from Kaggle

* Install Kaggle package to fetch data

In [5]:
! pip3 install kaggle==1.5.12

* In order to fetch the data you will need to have an authentication token from Kaggle available in your workspace directory, it takes the form of a file named 'kaggle.json'.
* In case you don't have it available, please follow the following steps:
* Log into your existing Kaggle account or create a new account and log in.
* Once you are logged in, on the top right of the page, click on your user profile picture, then on “Settings” from the dropdown menu. This will take you to your account settings.
* Scroll down to the section of the page called API.
* Click Expire API Token to remove any previous tokens.
* To create a new token, click on the Create New API Token button. It will generate a fresh authentication token and will download a kaggle.json file to your machine.
* Now transfer the file to your working directory and make sure it is named correctly.
* Once you have the file available run the cell below so that the token is recognized in the session.

In [7]:
import os
os.environ['KAGGLE_CONFIG_DIR'] = os.getcwd()
! chmod 600 kaggle.json

* The data set is located at the following URL: https://www.kaggle.com/datasets/codeinstitute/housing-prices-data
* Define the kaggle path to be the path that follows www.kaggle.com/datasets/ in the urls.
* Set the destination folder.
* Download the data.

In [8]:
KaggleDatasetPath = "codeinstitute/housing-prices-data"
DestinationFolder = "inputs/datasets/raw"
!kaggle datasets download -d {KaggleDatasetPath} -p {DestinationFolder} --force


  0%|                                               | 0.00/49.6k [00:00<?, ?B/s]
100%|██████████████████████████████████████| 49.6k/49.6k [00:00<00:00, 12.5MB/s]


* Unzip the downloaded file, delete the zip file and delete the kaggle.json file.

In [10]:
# ! unzip {DestinationFolder}/*.zip -d {DestinationFolder} \
#   && rm {DestinationFolder}/*.zip \
#   && rm kaggle.json

# Unzip the archive, overwriting any existing files
!unzip -o {DestinationFolder}/housing-prices-data.zip -d {DestinationFolder}

# Remove the downloaded ZIP file
!rm {DestinationFolder}/housing-prices-data.zip
# && rm kaggle.json


Archive:  inputs/datasets/raw/housing-prices-data.zip
  inflating: inputs/datasets/raw/house-metadata.txt  
  inflating: inputs/datasets/raw/house-price-20211124T154130Z-001/house-price/house_prices_records.csv  
  inflating: inputs/datasets/raw/house-price-20211124T154130Z-001/house-price/inherited_houses.csv  


In [11]:
!pip3 uninstall -y kaggle==1.5.12

Found existing installation: kaggle 1.5.12
Uninstalling kaggle-1.5.12:
  Successfully uninstalled kaggle-1.5.12


## Import packages & set environment variables

In [12]:
import pandas as pd
pd.options.display.max_columns = None
pd.options.display.max_rows = None

# Load and Inspect the House Price Records

* We can read the house_prices_records dataset csv file into a Pandas dataframe.

In [13]:
df = pd.read_csv(f"inputs/datasets/raw/house-price-20211124T154130Z-001/house-price/house_prices_records.csv")
print(df.shape)
df.head()

(1460, 24)


1stFlrSF  2ndFlrSF  BedroomAbvGr BsmtExposure  BsmtFinSF1 BsmtFinType1  \
0       856     854.0           3.0           No         706          GLQ   
1      1262       0.0           3.0           Gd         978          ALQ   
2       920     866.0           3.0           Mn         486          GLQ   
3       961       NaN           NaN           No         216          ALQ   
4      1145       NaN           4.0           Av         655          GLQ   

   BsmtUnfSF  EnclosedPorch  GarageArea GarageFinish  GarageYrBlt  GrLivArea  \
0        150            0.0         548          RFn       2003.0       1710   
1        284            NaN         460          RFn       1976.0       1262   
2        434            0.0         608          RFn       2001.0       1786   
3        540            NaN         642          Unf       1998.0       1717   
4        490            0.0         836          RFn       2000.0       2198   

  KitchenQual  LotArea  LotFrontage  MasVnrArea  OpenPorchSF  OverallCond  \
0          Gd     8450         65.0       196.0           61            5   
1          TA     9600         80.0         0.0            0            8   
2          Gd    11250         68.0       162.0           42            5   
3          Gd     9550         60.0         0.0           35            5   
4          Gd    14260         84.0       350.0           84            5   

   OverallQual  TotalBsmtSF  WoodDeckSF  YearBuilt  YearRemodAdd  SalePrice  
0            7          856         0.0       2003          2003     208500  
1            6         1262         NaN       1976          1976     181500  
2            7          920         NaN       2001          2002     223500  
3            7          756         NaN       1915          1970     140000  
4            8         1145         NaN       2000          2000     250000

## Dataframe Summary


* We are also able to read the dataframe * summary by calling the method .info() on the dataframe object, but the snippet below reads this output into it's own dataframe for readability purposes.

In [14]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 24 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   1stFlrSF       1460 non-null   int64  
 1   2ndFlrSF       1374 non-null   float64
 2   BedroomAbvGr   1361 non-null   float64
 3   BsmtExposure   1460 non-null   object 
 4   BsmtFinSF1     1460 non-null   int64  
 5   BsmtFinType1   1346 non-null   object 
 6   BsmtUnfSF      1460 non-null   int64  
 7   EnclosedPorch  136 non-null    float64
 8   GarageArea     1460 non-null   int64  
 9   GarageFinish   1298 non-null   object 
 10  GarageYrBlt    1379 non-null   float64
 11  GrLivArea      1460 non-null   int64  
 12  KitchenQual    1460 non-null   object 
 13  LotArea        1460 non-null   int64  
 14  LotFrontage    1201 non-null   float64
 15  MasVnrArea     1452 non-null   float64
 16  OpenPorchSF    1460 non-null   int64  
 17  OverallCond    1460 non-null   int64  
 18  OverallQ

* Check for missing values.

In [15]:
df.isnull().sum()

1stFlrSF            0
2ndFlrSF           86
BedroomAbvGr       99
BsmtExposure        0
BsmtFinSF1          0
BsmtFinType1      114
BsmtUnfSF           0
EnclosedPorch    1324
GarageArea          0
GarageFinish      162
GarageYrBlt        81
GrLivArea           0
KitchenQual         0
LotArea             0
LotFrontage       259
MasVnrArea          8
OpenPorchSF         0
OverallCond         0
OverallQual         0
TotalBsmtSF         0
WoodDeckSF       1305
YearBuilt           0
YearRemodAdd        0
SalePrice           0
dtype: int64

Check for duplicated entries.

In [16]:
df[df.duplicated(subset=None)]

Empty DataFrame
Columns: [1stFlrSF, 2ndFlrSF, BedroomAbvGr, BsmtExposure, BsmtFinSF1, BsmtFinType1, BsmtUnfSF, EnclosedPorch, GarageArea, GarageFinish, GarageYrBlt, GrLivArea, KitchenQual, LotArea, LotFrontage, MasVnrArea, OpenPorchSF, OverallCond, OverallQual, TotalBsmtSF, WoodDeckSF, YearBuilt, YearRemodAdd, SalePrice]
Index: []

* Check unique values in the columns with non-numeric data type

In [17]:
for col in df:
    if df[col].dtypes=='object':
        print(col, '-', df[col].unique())
    elif df[col].unique().size < 11:
        print(col, '-', df[col].unique().size)

BedroomAbvGr - 9
BsmtExposure - ['No' 'Gd' 'Mn' 'Av' 'None']
BsmtFinType1 - ['GLQ' 'ALQ' 'Unf' 'Rec' nan 'BLQ' 'None' 'LwQ']
GarageFinish - ['RFn' 'Unf' nan 'Fin' 'None']
KitchenQual - ['Gd' 'TA' 'Ex' 'Fa']
OverallCond - 9
OverallQual - 10


## Load and Inspect Inherited House Records

* Read the inherited_houses dataset csv file into a Pandas dataframe.

In [18]:
df_inherited = pd.read_csv(f"inputs/datasets/raw/house-price-20211124T154130Z-001/house-price/inherited_houses.csv")
print(df_inherited.shape)
df_inherited

(4, 23)


1stFlrSF  2ndFlrSF  BedroomAbvGr BsmtExposure  BsmtFinSF1 BsmtFinType1  \
0       896         0             2           No       468.0          Rec   
1      1329         0             3           No       923.0          ALQ   
2       928       701             3           No       791.0          GLQ   
3       926       678             3           No       602.0          GLQ   

   BsmtUnfSF  EnclosedPorch  GarageArea GarageFinish  GarageYrBlt  GrLivArea  \
0      270.0              0       730.0          Unf       1961.0        896   
1      406.0              0       312.0          Unf       1958.0       1329   
2      137.0              0       482.0          Fin       1997.0       1629   
3      324.0              0       470.0          Fin       1998.0       1604   

  KitchenQual  LotArea  LotFrontage  MasVnrArea  OpenPorchSF  OverallCond  \
0          TA    11622         80.0         0.0            0            6   
1          Gd    14267         81.0       108.0           36            6   
2          TA    13830         74.0         0.0           34            5   
3          Gd     9978         78.0        20.0           36            6   

   OverallQual  TotalBsmtSF  WoodDeckSF  YearBuilt  YearRemodAdd  
0            5        882.0         140       1961          1961  
1            6       1329.0         393       1958          1958  
2            5        928.0         212       1997          1998  
3            6        926.0         360       1998          1998

## Preliminary assessment of data

* The data set has 1460 entries/rows and 24 columns.
* The columns are a mix of data types, integers, floats, and objects.
* 9 columns are missing data to various degrees.
* 4 columns have non-numerical entries and are thus categorical data.
* 3 further columns have only a small number of unique numerical entries, suggesting that they may also be categorical values.
* Further investigation and Data Cleaning are suggested for the next Notebook.
* There is a difference between the house price and inherited house datasets.t* pricesprice is only found in the house price dataset in integerf* Further investigation and Data Cleaning are suggested for the next Notebook.ur data.


* Save the data file in a local folder named 'collection' under 'outputs/datasets/' and push it to the repository.

In [19]:
import os
try:
  os.makedirs(name='outputs/datasets/collection') # create a folder for the data output
except Exception as e:
  print(e)

df.to_csv(f"outputs/datasets/collection/house_prices_records.csv",index=False)

# Push files to Repo

* If you do not need to push files to Repo, you may replace this section with "Conclusions and Next Steps" and state your conclusions and next steps.